src/openloco/station.cpp

```diff
@@ -92,14 +230,304 @@ namespace openloco
+                            {
+                                cargoSearchState.filter(cargoSearchState.filter() | (1 << obj->var_2C));
+                            }
+                            else if (obj->flags & road_station_flags::freight)
+                            {
+                                cargoSearchState.filter(cargoSearchState.filter() | ~(1 << obj->var_2C));
+                            }
+                        }
+                        else
+                        {
+                            cargoSearchState.filter(~0);
+                        }
+                        break;
+                    }
+                }
+            }
+        }
+
+        if (cargoSearchState.filter() == 0)
+        {
+            cargoSearchState.filter(~0);
+        }
+
+        for (tile_coord_t ty = 0; ty < map_columns; ty++)
+        {
+            for (tile_coord_t tx = 0; tx < map_rows; tx++)
+            {
+                if (cargoSearchState.mapHas2(tx, ty))
+                {
+                    auto pos = map_pos(tx * tile_size, ty * tile_size);
+                    auto tile = tilemgr::get(pos);
+
+                    for (auto& el : tile)
+                    {
+                        if (!el.is_flag_4())
+                        {
+                            switch (el.type())
+                            {
+                                case element_type::industry:
+                                {
+                                    auto industryEl = el.as_industry();
+                                    auto industry = industryEl->industry();
+
+                                    if (industry != nullptr && industry->under_construction == 0xFF)
+                                    {
+                                        auto obj = industry->object();
+
+                                        if (obj != nullptr)
+                                        {
+                                            for (auto cargoId : obj->required_cargo_type)
+                                            {
+                                                if (cargoId != 0xFF && (cargoSearchState.filter() & (1 << cargoId)))
+                                                {
+                                                    cargoSearchState.addScore(cargoId, 8);
+                                                    cargoSearchState.setIndustry(cargoId, industry->id());
+                                                }
+                                            }
+
+                                            for (auto cargoId : obj->produced_cargo_type)
+                                            {
+                                                if (cargoId != 0xFF && (cargoSearchState.filter() & (1 << cargoId)))
+                                                {
+                                                    cargoSearchState.dword_112C710(cargoSearchState.dword_112C710() | (1 << cargoId));
```

Comment:
I think it would make more sense to have a function that worked like this `.addProducedCargoType(cargoId);`